### PR TITLE
Auto inject missing AMP boilerplate

### PR DESCRIPTION
**Pull request by sebastianbenz**
```
AmpOptimizer.transformHtml('<h1>Hello World</h1>', {
  canonical: 'canonical.html',
  title: 'the title',
})
```

will generate

```
<!DOCTYPE html>
<html amp="">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width,minimum-scale=1"><noscript><style amp-boilerplate="">body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript><style amp-boilerplate="">body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style>
  <script async="" src="https://cdn.ampproject.org/v0.js"></script>
  <link rel="canonical" href="canonical.html">
  <title>the title</title>
</head>
<body>
  <h1>Hello World</h1>
</body>
</html>
```

### CODE REVIEW

File: packages/optimizer/lib/transformers/AutoAddBoilerplate.js [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const {AMP_FORMATS, AMP_TAGS} = require('../AmpConstants');
+
+const DEFAULT_FORMAT = 'AMP';
+
+const BOILERPLATES = {
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/AutoAddBoilerplate.js]: I'll solve that problem when I get there...

File: packages/optimizer/lib/transformers/AutoAddBoilerplate.js [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+ * This transformer will only add missing or fix invalid boilerplate code. However, it won't remove invalid elements.
+ *
+ * This transformer supports the following option:
+ *
+ * - `format: [AMP|AMP4EMAIL|AMP4ADS]` - specifies the AMP format. Defaults to `AMP`.
+ * - `autoAddBoilerplate: [true|false]` - set to `false` to disable auto adding the boilerplate.
+ */
+class AutoAddBoilerplate {
+  constructor(config) {
+    this.enabled = config.autoAddBoilerplate !== false;
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/AutoAddBoilerplate.js]: I prefer false as it's more explicit

File: packages/optimizer/lib/transformers/AutoAddBoilerplate.js [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    // Add the doctype if none is present
+    let doctype = tree.root.firstChildByTag('!doctype');
+    if (!doctype) {
+      doctype = tree.createDocType('html');
+      tree.root.insertBefore(doctype, tree.root.firstChild);
+    }
+    const html = tree.root.firstChildByTag('html');
+
+    // Mark as AMP
+    if (!Object.keys(html.attribs).some((a) => AMP_TAGS.includes(a))) {
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/AutoAddBoilerplate.js]: Why would it add `amp` in that `amp4email` case?

Format auto-detect is not possible as no html tag might be given initially.

File: packages/optimizer/lib/transformers/AutoAddBoilerplate.js [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+  }
+
+  /**
+   * @private
+   */
+  matchSpec(matcher, node) {
+    if (matcher.tagName !== node.tagName) {
+      return false;
+    }
+    if (!matcher.attribs) {
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/AutoAddBoilerplate.js]: attribs might be null

File: packages/optimizer/lib/transformers/AutoAddBoilerplate.js [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+      }
+      node = node.nextSibling;
+    }
+
+    if (boilerplateRules.size === 0) {
+      return;
+    }
+
+    // Setup params (in case they're needed)
+    params.canonical = params.canonical || '.';
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/AutoAddBoilerplate.js]: moved to top, also added a warning

File: packages/optimizer/lib/transformers/AutoAddBoilerplate.js [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        boilerplateRules.forEach((spec) => {
+          if (this.matchSpec(spec.matcher, node)) {
+            boilerplateRules.delete(spec);
+          }
+        });
+      }
+      node = node.nextSibling;
+    }
+
+    if (boilerplateRules.size === 0) {
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/AutoAddBoilerplate.js]: done.

File: packages/optimizer/lib/transformers/AutoAddBoilerplate.js [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * @private
+   */
+  addNode(tree, node, matcher, params) {
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/AutoAddBoilerplate.js]: done. also renamed matcher to nodeSpec

File: packages/optimizer/lib/transformers/AutoAddBoilerplate.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (fstanis) [reply packages/optimizer/lib/transformers/AutoAddBoilerplate.js]: Whoops, I misread. Still, this means `<html amp4email>` will stay `<html amp4email>`, correct?

File: packages/optimizer/lib/transformers/AutoAddBoilerplate.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (fstanis) [reply packages/optimizer/lib/transformers/AutoAddBoilerplate.js]: `matcher.attribs = matcher.attribs || {}` then?

**Comment by fstanis** [comment]
> For me the boilerplate is everything you need to add for AMP by default. I know it's documented differently...

Right, it's unfortunate, but now your module's name will confuse anyone familiar with AMP. I strongly suggest using anything other than "boilerplate", e.g. "MandatoryTags".

> Moving these would make things harder to understand.

Strongly disagree: right now, all of your methods are unrelated to the class they belong to and either read or mutate from a plain object. This is a clear indicator that plain object should be a class. If you really don't want to do this, at least move them to be simple functions outside of the class. Currently, it makes no sense for matcher-related methods to be inside the transformer class, as they don't deal with the state of that class at all.

**Comment by ithinkihaveacat** [comment]
[pandoc](https://pandoc.org) calls the auto-inject header and footer mode "[standalone](https://pandoc.org/MANUAL.html#templates)".

**Comment by sebastianbenz** [comment]
Thanks both for the feedback.

Renamed the transformer to `AddMandatoryTags`. 

Re: refactoring the transformer -  I simply don't know yet how to best evolve the matchers. I see different options and will wait for future changes before deciding on how to best fix the current version.

**Pull request merged**